o4/o4.py

```diff
@@ -1033,6 +1034,100 @@ def man(cmd):
               f' full verification.')
 
 
+def o4_status(changelist, depot):
+    from subprocess import run, PIPE
+    from pprint import pprint
+
+    o4bin = find_o4bin()
+    manibin = os.path.join(os.path.dirname(o4bin), 'manifold')
+    here = os.getcwd()
+    depot = depot.rstrip('.')
+
+    print(f"*** INFO: o4 status {here}")
+    if not os.path.isdir('.o4'):
+        print("Never synced with o4.")
+        return
+    try:
+        cur = int(open('.o4/changelist').read().strip())
+    except ValueError:
+        print("*** WARNING: .o4/changelist is invalid.")
+        cur = None
+    except FileNotFoundError:
+        print("*** WARNING: .o4/changelist is missing.")
+        cur = None
+    if cur is None:
+        cur, fname = get_fstat_cache(changelist)
+    if cur is None:
+        print("*** ERROR: Current changelist could not be determined.")
+        return
+
+    print(f"Current changelist: {cur:,d}")
+    print(f"  - HEAD is {changelist:,d} (+{changelist-cur:,d})")
+
+    keep_case = f'| {o4bin} keep --case ' if sys.platform == 'darwin' else ''
+    cmd = f"{o4bin} fstat .@{cur} {keep_case}| {manibin} o4 drop --checksum"
+    print(f"Checksumming ({cmd}).")
+    print("Please be patient...")
+    res = run(cmd, stdout=PIPE, shell=True, universal_newlines=True)
+    crcs = {
+        f[F_PATH]: f
+        for f in sorted((fstat_split(f) for f in res.stdout.splitlines()),
+                        key=lambda x: x[F_PATH])
+    }
+
+    has_open = {f['depotFile'].replace(depot, ''): f for f in Pyforce('opened', '...')}
+
+    # [{
+    #     'code': 'stat',
+    #     'depotFile': '//sandbox/t/a',
+    #     'clientFile': '//pbergen-ltm-blt/sandbox/t/a',
+    #     'movedFile': '//sandbox/t/b',
+    #     'rev': '1',
+    #     'haveRev': '2',
+    #     'action': 'move/delete',
+    #     'change': 'default',
+    #     'type': 'text',
+    #     'user': 'pbergen',
+    #     'client': 'pbergen-ltm-blt'
+    # }, {
+    naughty = set(crcs.keys()) - set(has_open.keys())
+    renamed = {
+        k: v['movedFile'].replace(depot, '')
+        for k, v in has_open.items() if v.get('action') == 'move/delete'
+    }
+    missing = {f for f in crcs.keys() if f not in renamed and not os.path.exists(f)}
+    all_fnames = set(list(has_open.keys()) + list(crcs.keys()))
+
+    if all_fnames:
+        print("\nFiles with local modifications:")
+        print(" (!=Mismatch A=Added D=Deleted O=Open R=Renamed)\n")
+
+    for f in sorted(all_fnames):
+        ho = has_open.get(f, {})
+        if ho.get('action') == 'move/add':
+            continue
+
+        n = '!' if f in naughty else ' '
+        m = ' '
+        if f in missing:
+            m = 'D'
+        if f in crcs:
+            m = 'M'
```

Comment:
M isn't in the legend

Reply:
I noticed. There’s another following pr adding that. I was naughty and just merged it